### PR TITLE
Create and update transaction

### LINKED ISSUE #41: Receipt Table Add Column
- [x] add receipt_unique_id fields with varchar(50)

details refer discussion on #39

### LINKED ISSUE #49: createTransaction
- Endpoint URI: */transactions*
- Http Method: *POST*
- API Logic: 
    - event with not CANCELLED status
    - EXPENSE Type
        - receiptDTO cannot empty
        - create a claim record
- Progress
    - [x] development
    - [x] integration Test

### LINKED ISSUE #50: Delete Transactions
- Endpoint URI: */transactions*
- Http Method: *PUT*
- API Logic: 
    - update transaction status
- Remarks
    - claim record will not be remove
- Progress
    - [x] development
    - [x] integration Test

### LINKED ISSUE #53: Transaction Authorization
PreAuthorize Check for Transaction Resource
- [x] createTransaction: ROLE_ADMIN and Event Crew
- [x] updateTransaction: ROLE_ADMINand Event Head
- [x] getAllTransactions: ROLE_ADMIN
- [x] getAllTransactionsByEventId: ROLE_ADMIN and Event Crew

Unit Test on authorized role:
- [x] createTransaction
- [x] updateTransaction
- [x] getAllTransactions
- [x] getAllTransactionsByEventId

### PR DESCRIPTION
**Pull request by HowardYaw**
resolve #49 
resolve #41 
resolve #50 
resolve #53 
refer discussion on #48

### CODE REVIEW

**Comment by HowardYaw** [comment]
resolve #53  in 2a9f1bff27165469d2d75fe3bce854c742d92e64

File: src/test/java/com/thirdcc/webapp/web/rest/TransactionResourceIT.java [hunk 1]
```diff
@@ -141,58 +159,253 @@ public static Event createEventEntity() {
...
+        assertThat(receiptList).hasSize(receiptDatabaseSizeBeforeCreate + 1);
+        Receipt testReceipt = receiptList.get(receiptList.size() - 1);
+
+        // Validate the Transaction in the database
+        List<Transaction> transactionList = transactionRepository.findAll();
+        assertThat(transactionList).hasSize(databaseSizeBeforeCreate + 1);
+        Transaction testTransaction = transactionList.get(transactionList.size() - 1);
+        assertThat(testTransaction.getEventId()).isEqualTo(transaction.getEventId());
+        assertThat(testTransaction.getReceiptId()).isEqualTo(testReceipt.getId());
+        assertThat(testTransaction.getType()).isEqualTo(transaction.getType());
```
Comment (kachun333) [inline src/test/java/com/thirdcc/webapp/web/rest/TransactionResourceIT.java]: why don't write
```
 assertThat(testTransaction.getType()).isEqualTo(TransactionType.EXPENSE);
```

if you write like this, the test may produce false positive 😞

File: src/test/java/com/thirdcc/webapp/web/rest/TransactionResourceIT.java [hunk 2]
```diff
@@ -141,58 +159,253 @@ public static Event createEventEntity() {
...
+        // Validate Receipt Table
+        List<Receipt> receiptList = receiptRepository.findAll();
+        assertThat(receiptList).hasSize(receiptDatabaseSizeBeforeCreate + 1);
+        Receipt testReceipt = receiptList.get(receiptList.size() - 1);
+
+        // Validate the Transaction in the database
+        List<Transaction> transactionList = transactionRepository.findAll();
+        assertThat(transactionList).hasSize(databaseSizeBeforeCreate + 1);
+        Transaction testTransaction = transactionList.get(transactionList.size() - 1);
+        assertThat(testTransaction.getEventId()).isEqualTo(transaction.getEventId());
```
Comment (kachun333) [inline src/test/java/com/thirdcc/webapp/web/rest/TransactionResourceIT.java]: why don't write
```
assertThat(testTransaction.getEventId()).isEqualTo(savedEvent.getId());
```
if you write like this, the test may produce false positive 😞

File: src/test/java/com/thirdcc/webapp/web/rest/TransactionResourceIT.java [hunk 3]
```diff
@@ -245,100 +456,98 @@ public void getAllTransactionByEventId_WithCancelledEvent_ShouldThrow400() throw
...
+            .contentType(TestUtil.APPLICATION_JSON_UTF8)
+            .content(TestUtil.convertObjectToJsonBytes(transactionDTO)))
+            .andExpect(status().isOk());
+
+        // Validate the Transaction in the database
+        List<Transaction> transactionList = transactionRepository.findAll();
+        assertThat(transactionList).hasSize(databaseSizeBeforeUpdate);
+        Transaction testTransaction = transactionList.get(transactionList.size() - 1);
+        assertThat(testTransaction.getEventId()).isEqualTo(transaction.getEventId());
+        assertThat(testTransaction.getType()).isEqualTo(transaction.getType());
```
Comment (kachun333) [inline src/test/java/com/thirdcc/webapp/web/rest/TransactionResourceIT.java]: same thing. false positive may occur

**Review by kachun333 — CHANGES_REQUESTED** [review]

**Comment by kachun333** [comment]
please write test cases that test only authorized users able to access the endpoint.

e.g.
a user which don't have admin_role and not event head, access the endpoint should throw 403

**Comment by kachun333** [comment]
I believe #50 is not resolved yet.

I don't see any endpoint that changes the transaction to `cancelled` status

**Comment by HowardYaw** [comment]
@kachun333  can't we just use the update transaction API to change the transaction to CANCELLED?

**Comment by HowardYaw** [comment]
Update transaction can only update details and status

File: src/test/java/com/thirdcc/webapp/web/rest/TransactionResourceIT.java (hunk 1 above)
Comment (HowardYaw) [reply src/test/java/com/thirdcc/webapp/web/rest/TransactionResourceIT.java]: Resolved in 4104de2d993ff69fd961084fe0c4b0424b4afd30

File: src/test/java/com/thirdcc/webapp/web/rest/TransactionResourceIT.java (hunk 3 above)
Comment (HowardYaw) [reply src/test/java/com/thirdcc/webapp/web/rest/TransactionResourceIT.java]: Resolved in 4104de2d993ff69fd961084fe0c4b0424b4afd30

File: src/test/java/com/thirdcc/webapp/web/rest/TransactionResourceIT.java (hunk 2 above)
Comment (HowardYaw) [reply src/test/java/com/thirdcc/webapp/web/rest/TransactionResourceIT.java]: Resolved in 4104de2d993ff69fd961084fe0c4b0424b4afd30

**Review by LUXIANZE — APPROVED** [review]

**Review by CarlosSia — APPROVED** [review]

File: src/main/java/com/thirdcc/webapp/authorization/ManagementTeamSecurityExpression.java [hunk 4]
```diff
@@ -76,6 +71,14 @@ public boolean isEventHead(Long eventId) {
         return eventCrew != null && eventCrew.getRole().equals(EventCrewRole.HEAD);
     }
 
+    public boolean hasRoleAdminOrIsEventHead(Long eventId) {
+        if (getUserAuthRole().contains(AuthoritiesConstants.ADMIN)) {
+            return true;
+        } else {
+            return isEventHead(eventId);
+        }
```
Comment (kachun333) [inline src/main/java/com/thirdcc/webapp/authorization/ManagementTeamSecurityExpression.java:79]: @HowardYaw  can shorten this to 
```
return (getUserAuthRole().contains(AuthoritiesConstants.ADMIN)) || isEventHead(eventId)
```

**Review by kachun333 — APPROVED** [review]
